shared/util/quit-helper.desktop.js

```diff
@@ -55,8 +55,20 @@ export function executeActions (actions: Array<Action>) {
   }
 }
 
-export function setupExecuteActionsListener () {
+type appType = {
+  once: (name: string, callback:
+    (event: {preventDefault: () => void}) => void
+  ) => void
+}
+export function setupExecuteActionsListener (app: appType) {
   ipcMain.on('executeActions', (event, actions) => {
     executeActions(actions)
   })
+
+  // quit through dock. only listen once
+  app.once('before-quit', event => {
+    console.log('Quit through dock')
+    event.preventDefault()
+    quit()
```

Comment:
this is complicated because of the once logic. the other quit things will cause beforeQuitEvent to fire so i think there'll be a bunch of logic to make this work if we do this